lib/gc/Transforms/GPU/GpuTilingAndFusion.cpp

```diff
@@ -0,0 +1,481 @@
+      }
+      return WalkResult::skip();
+    });
+  }
+
+  static void tileForallOp(OpRewriter &rw, ForallOp op, int64_t wgSize) {
+    rw.loc = op.getLoc();
+    rw.setInsertionPoint(op);
+    OpFoldResult zero{rw.getConstant(0)};
+    OpFoldResult one{rw.getConstant(1)};
+    auto innerSteps = op.getMixedStep();
+    auto outerBounds = op.getMixedUpperBound();
+    SmallVector<OpFoldResult> outerSteps;
+    auto count = innerSteps.size();
+
+    { // Calculate outer steps
+      SmallVector<int64_t> tiles;
+      tiles.reserve(count);
+      for (auto s : innerSteps) {
+        if (auto v = getConstantIntValue(s)) {
+          tiles.emplace_back(*v);
+        } else {
+          // TODO: Add support for dynamic sizes
+          tiles.emplace_back(32);
+        }
+      }
+      adjustTiles(wgSize, tiles);
+      outerSteps.reserve(count);
+      for (auto [s, b, t] : zip(innerSteps, outerBounds, tiles)) {
+        if (auto sv = getConstantIntValue(s)) {
+          auto step = *sv * t;
+          if (auto bv = getConstantIntValue(b)) {
+            step = std::min(step, *bv);
+          }
+          outerSteps.emplace_back(rw.getConstant(step));
+        } else {
+          outerSteps.emplace_back(
+              rw.create<arith::MulIOp>(s.get<Value>(), rw.getConstant(t)));
+        }
+      }
+    }
+
+    auto outerLoop =
+        rw.create<ForallOp>(op.getMixedLowerBound(), outerBounds, outerSteps,
+                            op.getOutputs(), std::nullopt);
+    rw.setInsertionPointToStart(outerLoop.getBody());
+    SmallVector<OpFoldResult> innerBounds;
+    SmallVector<Range> ranges;
+
+    {
+      auto idxType = rw.getIndexType();
+      auto ctx = rw.getContext();
+      auto minMap = AffineMap::get(
+          /*dimCount=*/3, /*symbolCount=*/0,
+          {getAffineDimExpr(0, ctx),
+           getAffineDimExpr(1, ctx) - getAffineDimExpr(2, ctx)},
+          rw.getContext());
+      innerBounds.reserve(count);
+      ranges.reserve(count);
+      for (auto [i, u, s] : zip(outerLoop.getInductionVars(),
+                                outerLoop.getMixedUpperBound(), outerSteps)) {
+        OpFoldResult iub;
+        auto cu = getConstantIntValue(u);
+        auto cs = getConstantIntValue(s);
+        if (cu && cs && (*cu % *cs == 0)) {
+          iub = s;
+        } else {
+          Value vub = cu ? rw.getConstant(*cu) : u.get<Value>();
+          Value vs = cs ? rw.getConstant(*cs) : s.get<Value>();
+          iub = OpFoldResult(rw.create<affine::AffineMinOp>(
+              idxType, minMap, ValueRange{vs, vub, i}));
+        }
+        innerBounds.emplace_back(iub);
+        ranges.emplace_back(Range{i, iub, one});
+      }
+    }
+
+    SmallVector<Value> innerOutputs;
+    for (auto o : outerLoop.getRegionIterArgs()) {
+      innerOutputs.emplace_back(rw.create<tensor::ExtractSliceOp>(o, ranges));
+    }
+
+    auto innerLoop =
+        rw.create<ForallOp>(SmallVector<OpFoldResult>(count, zero), innerBounds,
```

Comment:
Why create the forallop manaually instead of reusing the `tileUsingSCF`?

Reply:
`tileUsingSCF()` tiles an op that implements the `TilingInterface`. ForallOp is not the case.